File: android/src/main/java/io/boxo/sdk/capacitor/AppboxoPlugin.kt
package io.boxo.sdk.capacitor

import android.os.Handler
import android.os.Looper
import com.appboxo.data.models.MiniappData
import com.appboxo.js.params.CustomEvent
import com.appboxo.js.params.PaymentData
import com.appboxo.sdk.Appboxo
import com.appboxo.sdk.Config
import com.appboxo.sdk.Miniapp
import com.appboxo.sdk.MiniappConfig
import com.appboxo.sdk.MiniappListCallback
import com.appboxo.ui.main.AppboxoActivity
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "Appboxo")
class AppboxoPlugin : Plugin(), Miniapp.LifecycleListener,
    Miniapp.CustomEventListener, Miniapp.AuthListener,
    Miniapp.PaymentEventListener{

    private var handler: Handler? = null

    override fun load() {
        super.load()
        Appboxo.init(bridge.activity.application)
        handler = Handler(Looper.getMainLooper())
    }

    @PluginMethod
    fun setConfig(call: PluginCall) {
        val clientId = call.getString("clientId")!!
        val userId = call.getString("userId")?:""
        val sandboxMode = call.getBoolean("sandboxMde", false)!!
        val enableMultitaskMode = call.getBoolean("enableMultitaskMode", false)!!
        val theme = call.getString("theme", "system")!!
        val isDebug = call.getBoolean("isDebug", false)!!
        val showPermissionsPage = call.getBoolean("showPermissionsPage", true)!!
        val showClearCache = call.getBoolean("showClearCache", true)!!
        val globalTheme: Config.Theme = when (theme) {
            "light" -> Config.Theme.LIGHT
            "dark" -> Config.Theme.DARK
            else -> Config.Theme.SYSTEM
        }
        Appboxo
            .setConfig(
                Config.Builder()
                    .setClientId(clientId)
                    .setUserId(userId)
                    .sandboxMode(sandboxMode)
                    .multitaskMode(enableMultitaskMode)
                    .setTheme(globalTheme)
                    .permissionsPage(showPermissionsPage)
                    .showClearCache(showClearCache)
                    .debug(isDebug)
                    .build()
            )
    }

    @PluginMethod
    fun openMiniapp(call: PluginCall) {
        val appId = call.getString("appId")!!
        val data = call.getObject("data")
        val theme = call.getString("theme")
        val extraUrlParams = call.getObject("extraUrlParams")
        val urlSuffix = call.getString("urlSuffix")
        val colors = call.getObject("colors")
        val miniapp: Miniapp = Appboxo.getMiniapp(appId)
            .setCustomEventListener(this)
            .setPaymentEventListener(this)
            .setAuthListener(this)
            .setLifecycleListener(this)
        if (data != null) miniapp.setData(data.toMap())
        val configBuilder = MiniappConfig.Builder()
        if (theme != null) {
            val miniappTheme: Config.Theme? = when (theme) {
                "light" -> Config.Theme.LIGHT
                "dark" -> Config.Theme.DARK
                "system" -> Config.Theme.SYSTEM
                else -> null
            }
            if (miniappTheme != null) {
                configBuilder.setTheme(miniappTheme)
            }
        }
        if (extraUrlParams != null) {
            val map: Map<String, Any> = extraUrlParams.toMap()
            val stringMap: MutableMap<String, String> = HashMap()
            for ((key, value) in map) stringMap[key] = value.toString()
            configBuilder.setExtraUrlParams(stringMap)
        }
        urlSuffix?.also { suffix -> configBuilder.setUrlSuffix(suffix) }
        if (colors != null) {
            configBuilder.setColors(
                colors.getString("primaryColor") ?: "",
                colors.getString("secondaryColor") ?: "",
                colors.getString("tertiaryColor") ?: "",
            )
        }
        miniapp.setConfig(configBuilder.build())
        miniapp.open(bridge.activity)
    }

    @PluginMethod
    fun setAuthCode(call: PluginCall) {
        val appId = call.getString("appId")!!
        val authCode = call.getString("authCode")!!
        Appboxo.getMiniapp(appId)
            .setAuthCode(authCode)
    }

    @PluginMethod
    fun closeMiniapp(call: PluginCall) {
        val appId = call.getString("appId")!!
        handler?.post { Appboxo.getExistingMiniapp(appId)?.close() }
    }

    override fun handle(
        miniAppActivity: AppboxoActivity,
        miniapp: Miniapp,
        customEvent: CustomEvent
    ) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        try {
            params.put("requestId", customEvent.requestId ?: 0)
            params.put("type", customEvent.type)
            params.put("errorType", customEvent.errorType)
            params.put("payload", customEvent.payload)
            sendEvent(CUSTOM_EVENTS_EVENT_NAME, params)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onLaunch(miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onLaunch")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onResume(miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onResume")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onPause(miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onPause")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onClose(miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onClose")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onError(miniapp: Miniapp, message: String) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onError")
        params.put("error", message)
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onUserInteraction(miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onUserInteraction")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    override fun onAuth(appboxoActivity: AppboxoActivity, miniapp: Miniapp) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        params.put("lifecycle", "onAuth")
        sendEvent(MINIAPP_LIFECYCLE_EVENT_NAME, params)
    }

    @PluginMethod
    fun sendCustomEvent(call: PluginCall) {
        try {
            val appId: String = call.getString("appId")!!
            val event = CustomEvent(
                call.getInt("requestId"),
                call.getString("type")!!,
                call.getString("errorType"),
                call.getObject("payload")?.toMap() ?: emptyMap()
            )
            handler?.post { Appboxo.getExistingMiniapp(appId)?.sendEvent(event) }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @PluginMethod
    fun sendPaymentEvent(call: PluginCall) {
        try {
            val appId: String = call.getString("appId")!!
            val data = PaymentData(
                call.getString("transactionToken") ?: "",
                call.getString("miniappOrderId") ?: "",
                call.getDouble("amount")?:0.0,
                call.getString("currency") ?: "",
                call.getString("status") ?: "",
                call.getString("hostappOrderId") ?: "",
                call.getObject("extraParams")?.toMap()
            )
            handler?.post { Appboxo.getExistingMiniapp(appId)?.sendPaymentResult(data) }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @PluginMethod
    fun getMiniapps(call: PluginCall) {
        Appboxo.getMiniapps(object : MiniappListCallback {
            override fun onSuccess(miniapps: List<MiniappData>) {
                try {
                    val list  = JSArray()
                    for (i in miniapps.indices) {
                        val miniappData: MiniappData = miniapps[i]
                        val data = JSObject()
                        data.put("appId", miniappData.appId)
                        data.put("name", miniappData.name)
                        data.put("category", miniappData.category)
                        data.put("logo", miniappData.logo)
                        data.put("description", miniappData.description)
                        list.put(data)
                    }
                    val params = JSObject()
                    params.put("miniapps", list)
                    call.resolve(params)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            override fun onFailure(e: Exception) {
                val result = JSObject()
                result.put("error", e.message)
                call.resolve(result)
            }
        })
    }

    @PluginMethod
    fun hideMiniapps() {
        Appboxo.hideMiniapps()
    }

    @PluginMethod
    fun logout() {
        Appboxo.logout()
    }

    private fun sendEvent(name: String, data: JSObject) {
        notifyListeners(name, data)
    }

    override fun handle(
        miniAppActivity: AppboxoActivity,
        miniapp: Miniapp,
        paymentData: PaymentData
    ) {
        val params = JSObject()
        params.put("appId", miniapp.appId)
        try {
            params.put("transactionToken", paymentData.transactionToken)
            params.put("miniappOrderId", paymentData.miniappOrderId)
            params.put("amount", paymentData.amount)
            params.put("currency", paymentData.currency)
            params.put("status", paymentData.status)
            params.put("hostappOrderId", paymentData.hostappOrderId)
            if (paymentData.extraParams != null) {
                params.put("extraParams", paymentData.extraParams)
            }
            sendEvent(PAYMENT_EVENTS_EVENT_NAME, params)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    companion object {
        private const val CUSTOM_EVENTS_EVENT_NAME = "custom_event"
        private const val PAYMENT_EVENTS_EVENT_NAME = "payment_event"
        private const val MINIAPP_LIFECYCLE_EVENT_NAME = "miniapp_lifecycle"

        fun JSObject.toMap(): Map<String, Any> {
            val map: MutableMap<String, Any> = HashMap()
            val iterator: Iterator<String> = this.keys()

            while (iterator.hasNext()) {
                val key = iterator.next()
                var value: Any = this.get(key)

                if (value is JSObject) {
                    value = value.toMap()
                }
                if (value is JSArray) {
                    value = value.toList<Any>()
                }
                map[key] = value
            }

            return map
        }
    }
}

File: ios/Sources/AppboxoPlugin/AppboxoPlugin.swift
import Foundation
import Capacitor
import AppBoxoSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(AppboxoPlugin)
public class AppboxoPlugin: CAPPlugin, CAPBridgedPlugin {
    private let CUSTOM_EVENT = "custom_event"
    private let PAYMENT_EVENT = "payment_event"
    private let MINIAPP_LIFECYCLE_EVENT = "miniapp_lifecycle"

    public let identifier = "AppboxoPlugin"
    public let jsName = "Appboxo"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setConfig", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openMiniapp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAuthCode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "closeMiniapp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendCustomEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendPaymentEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMiniapps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideMiniapps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise)
    ]

    @objc func setConfig(_ call: CAPPluginCall) {
        let clientId = call.getString("clientId") ?? ""
        let userId = call.getString("userId") ?? ""
        let sandboxMode = call.getBool("sandboxMde", false)
        let theme = call.getString("theme", "system")
        let showPermissionsPage = call.getBool("showPermissionsPage", true)
        let showClearCache = call.getBool("showClearCache", true)
        var globalTheme : Theme = .System
        switch (theme) {
        case "dark":
            globalTheme = .Dark
        case "light":
            globalTheme = .Light
        default:
            globalTheme = .System
        }

        let config = Config(clientId: clientId, theme: globalTheme)
        config.sandboxMode = sandboxMode
        config.permissionsPage = showPermissionsPage
        config.showClearCache = showClearCache
        config.setUserId(id: userId)

        Appboxo.shared.setConfig(config: config)
    }

    @objc func openMiniapp(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""
        let data = call.getObject("data")?.toMap() ?? nil
        let theme = call.getString("theme")
        let extraUrlParams = call.getObject("extraUrlParams")?.toMap() ?? nil
        let colors = call.getObject("colors")?.toMap() ?? nil
        let enableSplash = call.getBool("enableSplash")

        let miniApp = Appboxo.shared.getMiniapp(appId: appId)
        miniApp.setData(data: data)
        miniApp.delegate = self

        let miniappConfig = MiniappConfig()
        if let enableSplash = enableSplash {
            miniappConfig.enableSplash(isSplashEnabled: enableSplash)
        }

        if let theme = theme {
            switch theme {
            case "dark":
                miniappConfig.setTheme(theme: .Dark)
            case "light":
                miniappConfig.setTheme(theme: .Light)
            case "system":
                miniappConfig.setTheme(theme: .System)
            default:
                break
            }
        }

        if let extraUrlParams = extraUrlParams as? [String : String] {
            miniappConfig.setExtraParams(extraParams: extraUrlParams)
        }

        if let colors = colors {
            miniappConfig.setColor(color: MiniappColor(primary: colors["primary_color"] as? String ?? "", secondary: colors["secondary_color"] as? String ?? "", tertiary: colors["tertiary_color"] as? String ?? ""))
        }

        miniApp.setConfig(config: miniappConfig)

        guard let viewController = bridge?.viewController else { return }
        DispatchQueue.main.async {
            miniApp.open(viewController: viewController)
        }

    }

    @objc func setAuthCode(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""
        let authCode = call.getString("authCode") ?? ""

        DispatchQueue.main.async {
            Appboxo.shared.getMiniapp(appId: appId).setAuthCode(authCode: authCode)
        }
    }

    @objc func closeMiniapp(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""

        if let miniapp = Appboxo.shared.getExistingMiniapp(appId: appId) {
            DispatchQueue.main.async {
                miniapp.close()
            }
        }
    }

    @objc func sendCustomEvent(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""

        let customEvent = CustomEvent()
        customEvent.requestId = call.getInt("requestId", 0)
        customEvent.type = call.getString("type") ?? ""
        customEvent.errorType = call.getString("errorType") ?? ""
        customEvent.payload = call.getObject("payload")?.toMap()

        DispatchQueue.main.async {
            Appboxo.shared.getMiniapp(appId: appId).sendCustomEvent(customEvent: customEvent)
        }
    }

    @objc func sendPaymentEvent(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""

        let paymentData = PaymentData()
        paymentData.transactionToken = call.getString("transactionToken") ?? ""
        paymentData.miniappOrderId = call.getString("miniappOrderId") ?? ""
        paymentData.amount = call.getDouble("amount", 0.0)
        paymentData.currency = call.getString("currency") ?? ""
        paymentData.status = call.getString("status") ?? ""
        paymentData.hostappOrderId = call.getString("hostappOrderId") ?? ""
        paymentData.extraParams = call.getObject("extraParams")?.toMap()

        DispatchQueue.main.async {
            Appboxo.shared.getMiniapp(appId: appId).sendPaymentEvent(paymentData: paymentData)
        }
    }

    @objc func getMiniapps(_ call: CAPPluginCall) {
        Appboxo.shared.getMiniapps { miniapps, error in
            if let error = error {
                call.reject(error)
            } else {
                var list = JSArray()

                miniapps.forEach { miniappData in
                    var data = JSObject()
                    data["appId"] = miniappData.appId
                    data["name"] = miniappData.name
                    data["category"] = miniappData.category
                    data["logo"] = miniappData.logo
                    data["description"] = miniappData.longDescription
                    list.append(data)
                }
                call.resolve(["miniapps" : list])
            }
        }
    }

    @objc func hideMiniapps(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Appboxo.shared.hideMiniapps()
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Appboxo.shared.logout()
        }
    }
}

extension AppboxoPlugin : MiniappDelegate {
    public func didReceivePaymentEvent(miniapp: Miniapp, paymentData: PaymentData) {
        let dict = [
            "appId" : miniapp.appId,
            "transactionToken" : paymentData.transactionToken,
            "miniappOrderId" : paymentData.miniappOrderId,
            "amount" : paymentData.amount,
            "currency" : paymentData.currency,
            "status" : paymentData.status,
            "hostappOrderId" : paymentData.hostappOrderId,
            "extraParams" : paymentData.extraParams ?? [String : Any]()
        ] as [String: Any]

        notifyListeners(PAYMENT_EVENT, data: dict)
    }

    public func didReceiveCustomEvent(miniapp: Miniapp, customEvent: CustomEvent) {
        let dict = [
            "appId" : miniapp.appId,
            "requestId" : customEvent.requestId,
            "type" : customEvent.type,
            "errorType" : customEvent.errorType,
            "payload" : customEvent.payload ?? [String : Any]()
        ] as [String: Any]

        notifyListeners(CUSTOM_EVENT, data: dict)
    }

    public func onLaunch(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onLaunch"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onResume(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onResume"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onPause(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onPause"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onClose(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onClose"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onError(miniapp: Miniapp, message: String) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onError",
            "error" : message
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onUserInteraction(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onUserInteraction"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }

    public func onAuth(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onAuth"
        ]

        notifyListeners(MINIAPP_LIFECYCLE_EVENT, data: dict)
    }
}

extension JSObject {
    func toMap() -> [String : Any] {
        var dict = [String : Any]()

        keys.forEach { key in
            if let object = self[key] as? JSObject {
                dict[key] = object.toMap()
            }

            dict[key] = self[key]
        }

        return dict
    }
}
